classes/class2/README.md

```diff
@@ -1 +1,309 @@
+
+Putting `()` next to the name of a function calls the function, so when we print `func1()` we get the return value of `func1`, which is `1`.
+
+`()` will actually always attempt to "call" the expression to its left, treating it like a function. So you can get errors like this:
+
+```node
+> 5()
+TypeError: number is not a function
+```
+
+A variable can also store a function declaration, so we could have done this instead:
+
+```node
+var func1 = function() {
+	return 1;
+};
+```
+
+The above example wouldn't have acted *quite* the same, though, which brings us to what makes JavaScript functions so special.
+
+### Functions are Objects
+
+The print statements above hinted at this: in JavaScript, functions are objects. So something like this is possible:
+
+###### functions.js
+```node
+function func1() {
+	var func2 = function() {
+		return 2
+	};
+	return func2;
+}
+```
+
+What will happen when we call `func1`?
+
+###### functions.js
+```node
+...
+console.log(func1());
+```
+
+```bash
+$ node functions.js
+[Function]
+```
+
+The function `func1` is a named function that returns the variable `func2`. We set `func2` equal to a function definition without a name — the function on the right hand side of this assignment is known as an **anonymous function**.
+
+So, `func1` returns a variable containing an anonymous function which returns `2`. How do we call the returned anonymous function?
+
+###### functions.js
+```node
+...
+console.log(func1()());
+```
+
+```bash
+$ node functions.js
+2
+```
+
+This is totally legal. It might help to break down the expression we just printed:
+
+```node
+func1()(); // first, func1 is called
+func2(); // it returns func2
+(function() { return 2; })(); // func2 holds an anonymous function
+2 // the anonymous function is called and returns 2
+```
+
+## Callbacks
+
+Function objects let us pass around pieces of code that can be run at arbitrary times.
+
+Here's a really common situation: A client application requests data from a URL, and wants to process the data when it arrives. There's no way to know how long it will take the data to arrive, and even if there was, we wouldn't want to just block execution and wait until we got the data. This would be **synchronous execution**, where code is executed one line at a time, in order. It would be great if we could request the data and go about our business until we received it — then we could run the code that depends on the data. **Asynchronous execution**, where code is executed out of order (often in response to events), allows this. Luckily, this behavior is the norm in JavaScript.
+
+**Callbacks** are functions that we provide to be run when an event occurs. Here's an example with real code:
```

Comment:
It's more like callbacks are called when the function you pass them to is ready, but this gets the right idea across regardless.